arangod/RocksDBEngine/RocksDBHotBackup.cpp

```diff
@@ -674,15 +663,30 @@ void RocksDBHotBackupRestore::execute() {
     if (good) {
       /// Step 2. initiate shutdown and restart with new data directory
       restoreExistingPath = getRocksDBPath();
-      restoreFailsafePath = rebuildPath(dirFailsafeString);
+      if (_saveCurrent) {
+        restoreFailsafePath = buildDirectoryPath(timepointToString(std::chrono::system_clock::now()),
+                                                                   "before_restore");
+        failsafeName = TRI_Basename(restoreFailsafePath.c_str());
+      } else {
+        failsafeName=dirFailsafeString;
+        if (0==failsafeName.compare(_directoryRestore)) {
+          failsafeName += ".1";
+        } // if
+        restoreFailsafePath = rebuildPath(failsafeName);
+      };
       clearPath(restoreFailsafePath);
 
       restartAction = new std::function<int()>();
       *restartAction = localRestoreAction;
-      application_features::ApplicationServer::server->beginShutdown();
+      startGlobalShutdown();
       _success = true;
-      _result.add(VPackValue(VPackValueType::Object));
-      _result.close();
+
+      try {
+        _result.add(VPackValue(VPackValueType::Object));
+        _result.add("previous", VPackValue(failsafeName));
+        _result.close();
+      } catch (...) {
```

Comment:
If this throws, the state of `_result` is unknown, i.e. invalid VPack. I don't know if it is a bad idea to send it anyways. As far as I know everything in ArangoDB assumes to have valid VPack, in particular the `toJson()` methods, which would crash because of a object that has not been closed. And so on ...

Maybe we should reset the builder `_result` in the `catch` case and fill it with none and report a internal server error?

Reply:
Agreed.  Doing that today ... and will see if there are similar code pieces that need same.

Reply:
One more comment, we need to check one thing: This might create a race between the shutdown and the sending out of the response. This has happened with the shutdown handler before. We might need to post a routine to the scheduler which waits for a few seconds before initiating the shutdown here.

Reply:
@neunhoef One second delay is checked in and available for review.